### PR TITLE
WIP: Implement po4a translation framework for the LinuxCNC documentation

### PR DESCRIPTION
**Pull request by petterreinholdtsen**
Handle both adoc and manual pages.  All texts are grouped into one POT file, as splitting them into groups did not work with the version of po4a I got.  This feature is implemented in the latest po4a edition and we should consider splitting into several POT files when this is available.

### CODE REVIEW

**Comment by petterreinholdtsen** [comment]
Note, migrating existing translations to PO files is still an unsolved issue.  This need to be done by someone who understand the language in question.

**Comment by petterreinholdtsen** [comment]
https://github.com/mquinson/po4a/commit/033673a7ca8be8540b3f592377e5eea94c1661ad is the po4a fix in question.

**Comment by smoe** [comment]
Is there any way for me to help? Sidenote - I think github is at fault not to offer me the option to send you a PR, anyway, please adopt
```
--- a/scripts/travis-install-build-deps.sh
+++ b/scripts/travis-install-build-deps.sh
@@ -16,3 +16,6 @@ if ! sudo apt-get -y build-dep . ; then
     exit 1
 fi
 sudo apt install -y lintian
+# checking validity of links - needs online access, hence not applicable
+# as part of the Debian package building and not a package build dependency
+sudo apt-get install -y w3c-linkchecker
```
to likely make the CI happy.

**Comment by petterreinholdtsen** [comment]
[Steffen Möller]
> Is there any way for me to help?

The migration currently hang on someone understanding Spanish, French
and Chinese, to avoid throwing away existing translations.  Do you
understand any of these?

The task is to make sure the original and translated documents have the
same structure, to allow the conversion script to match the strings and
place them in a PO file.  See docs/po/migrate-translation for the helper
script I created for this.

Without anyone able to read and match strings, I see no alternative to
throwing away the existing translations and start from scratch.

-- 
Happy hacking
Petter Reinholdtsen

**Comment by smoe** [comment]
>> [Steffen Möller]
>> Is there any way for me to help?
> The migration currently hang on someone understanding Spanish, French and Chinese, to avoid throwing away existing translations. Do you understand any of these? The task is to make sure the original and translated documents have the same structure, to allow the conversion script to match the strings and place them in a PO file. See docs/po/migrate-translation for the helper script I created for this. Without anyone able to read and match strings, I see no alternative to throwing away the existing translations and start from scratch.

Well, I have some general understanding of romance languages. I very much suggest to start from scratch, though. Just maybe give the existing translations a _XX_manual.adoc rename and request respective help from the community to adjust them.

Could you please prepare a README somewhere about what to do to start translating for a new language? And how to contribute to the existing translations? And then add a call for help for es, fr, vn and cn right in that Readme.

**Comment by smoe** [comment]
https://github.com/LinuxCNC/linuxcnc/runs/4155732618?check_suite_focus=true states the checklink was not found again. That is annoying. I checked and only saw the installation of lintian, but nof of w3c-linkchecker. I admit not to know if that is intended this way since you do not want arbitrary commits to invalidate their build tests, so maybe you have another idea where to introduce the installation of w3c-linkchecker.

**Comment by smoe** [comment]
Maybe you add an "echo '[ok]'" to the end of that script? The -e flag is already set.

**Comment by petterreinholdtsen** [comment]
[Steffen Möller]
> Maybe you add an "echo '[ok]'" to the end of that script? The -e flag
> is already set.

Asked on #linuxcnc-devel, and apparently the problem is really that
'master' is currently broken, and the problem is simply that I rebased
off a broken master.  We should wait for master to get fixed, I believe.

I'll have a look at making a README with some instructions.  Note, the
current scripts are just a stopgap solution, the operations should be
integrated into the makefiles before we merge into master.

-- 
Happy hacking
Petter Reinholdtsen

**Comment by silopolis** [comment]
Hey there!

Just discovered the asciidoc work and all done on documentation lately...
That's amazing!

Le mar. 9 nov. 2021 à 18:44, petterreinholdtsen ***@***.***>
a écrit :

> [Steffen Möller]
> > Is there any way for me to help?
>
> The migration currently hang on someone understanding Spanish, French
> and Chinese, to avoid throwing away existing translations. Do you
> understand any of these?
>
> The task is to make sure the original and translated documents have the
> same structure, to allow the conversion script to match the strings and
> place them in a PO file. See docs/po/migrate-translation for the helper
> script I created for this.
>
> Without anyone able to read and match strings, I see no alternative to
> throwing away the existing translations and start from scratch.


I'm french and would gladly help there if I can!

Could you point me exactly where and how I can join forces please?

Other comment to come in the translation rfc discussion...

Have a nice day
J

**Comment by smoe** [comment]
@petterreinholdtsen Please _vi -> _vn in https://github.com/LinuxCNC/linuxcnc/blob/8590a9de8fbdfc968506cc8dd41fa90bf4dfb55f/docs/src/Submakefile#L1 and https://github.com/LinuxCNC/linuxcnc/blob/8590a9de8fbdfc968506cc8dd41fa90bf4dfb55f/docs/src/Submakefile#L365 and https://github.com/LinuxCNC/linuxcnc/blob/8590a9de8fbdfc968506cc8dd41fa90bf4dfb55f/docs/src/Submakefile#L463 and https://github.com/LinuxCNC/linuxcnc/blob/8590a9de8fbdfc968506cc8dd41fa90bf4dfb55f/docs/src/Submakefile#L477 and whereever I may have missed it :)

@silopolis Petter's README is at https://github.com/LinuxCNC/linuxcnc/blob/8590a9de8fbdfc968506cc8dd41fa90bf4dfb55f/docs/README.adoc but I admit not to have followed it, yet. @petterreinholdtsen , you would not possibly feel like offering an spontanous web conference about it?

**Comment by petterreinholdtsen** [comment]
[Steffen Möller]
> @petterreinholdtsen Please _vi -> _vn in

Why?  'vi' is the code for the Vietnamese language, and
docs/html/gcode_vi.html is Vietnamese as far as I can tell,  while vn is
not in used according to
<URL: https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes >.

> @silopolis Petter's README is at
> https://github.com/LinuxCNC/linuxcnc/blob/8590a9de8fbdfc968506cc8dd41fa90bf4dfb55f/docs/README.adoc
> but I admit not to have followed it, yet. @petterreinholdtsen , you
> would not possibly feel like offering an spontanous web conference
> about it?

The README instructions are only partly updated, and mostly based on
work done earlier by Sebastian, so not yet ready for consumption.

I am happy to provide instructions via both IRC and a video conference.
Ping me on #linuxcnc-devel.

In short the task consist of reformatting the translated documents to
match the structure of the English original, and then convert to PO
files.  The script docs/po/migrate-translation try to do this migration,
so a prospective translator can run it and fix the translated files to
get the migration script working.  Once it is running without errors, I
can take over and do the migration.

-- 
Happy hacking
Petter Reinholdtsen

**Comment by smoe** [comment]
You have a point. This is about languages, not about countries. At least I learned something. Thanks!

**Comment by silopolis** [comment]
Le dim. 14 nov. 2021 à 19:30, petterreinholdtsen ***@***.***>
a écrit :

> @silopolis Petter's README is at
> >
> https://github.com/LinuxCNC/linuxcnc/blob/8590a9de8fbdfc968506cc8dd41fa90bf4dfb55f/docs/README.adoc
> > but I admit not to have followed it, yet. @petterreinholdtsen , you
> > would not possibly feel like offering an spontanous web conference
> > about it?
>
> The README instructions are only partly updated, and mostly based on
> work done earlier by Sebastian, so not yet ready for consumption.
>


I'll try to reach up asap with this...


I am happy to provide instructions via both IRC and a video conference.
> Ping me on #linuxcnc-devel.
>
> In short the task consist of reformatting the translated documents to
> match the structure of the English original, and then convert to PO
> files. The script docs/po/migrate-translation try to do this migration,
> so a prospective translator can run it and fix the translated files to
> get the migration script working. Once it is running without errors, I
> can take over and do the migration.


Give me a couple of days to wrap my head around this and get back to you...

**Comment by silopolis** [comment]
> In short the task consist of reformatting the translated documents to
> match the structure of the English original

Just to be sure I understand correctly, by "structure", you mean from the title and sections down to the paragraph/blank lines level? So that there's a "string to string" correspondance between the translated and english files? 

If so, I'll have another look, but from what I remember, chances are that for french at least it may well be better starting from scratch! In which case, if existing files in non english docs hold contents deserving to be preserved, then these should in turn be translated into english to sync every translations

**Comment by petterreinholdtsen** [comment]
[silopolis]
> Just to be sure I understand correctly, by "structure", you mean from
> the title and sections down to the paragraph/blank lines level? So
> that there's a "string to string" correspondance between the
> translated and english files?

Yes.  And if there is a mismatch, we can just add 'dummy' titles and
paragraphs to match up the real translations and trim away all 'dummy'
entries after the migration.

> If so, I'll have another look, but from what I remember, chances are
> that for french at least it may well be better starting from scratch!

OK.

> In which case, if existing files in non english docs hold contents
> deserving to be preserved, then these should in turn be translated
> into english to sync every translations

I agree.

-- 
Happy hacking
Petter Reinholdtsen

**Comment by petterreinholdtsen** [comment]
The migration to po4a is moving along.  The spanish PO file migration is done while the French is slowly moving forward.  Sadly no-one showed up to migrate the other languages, so the plan is to start from scratch there.

**Comment by smoe** [comment]
Should the changes to the English documentation not possibly be merged before since otherwise the .po files will not match as good and then some .po entries that are fine now would be lost in the next iteration?

**Comment by petterreinholdtsen** [comment]
[Steffen Möller]
> Should the changes to the English documentation not possibly be merged
> before since otherwise the .po files will not match as good and then
> some .po entries that are fine now would be lost in the next
> iteration?

Such changes will have to be handled outside this pull request.  I
suggest to handle it on Weblate during the translation review process.

I see the automatic tests fail with the new build rules.  I am unable to
see why.  Anyone got a clue to spare?

-- 
Happy hacking
Petter Reinholdtsen

**Comment by smoe** [comment]
The errors are shown in https://github.com/LinuxCNC/linuxcnc/runs/5265160863?check_suite_focus=true#step:5:7915 - some .po-generated(?) _fr files (and others) are using references that were fine in the original _fr file but with the auto-translated _fr files that are effectively still English, those references are different.

I locally started to work on this.

**Comment by smoe** [comment]
The file html/gcode_fr causes the errors (https://github.com/LinuxCNC/linuxcnc/runs/5265160863?check_suite_focus=true#step:5:7901) . This is (other than I initially thought) not autogenerated from the .adoc and dependent on the .po files but is edited manually. The file has stale references since the gcode/gcode_fr.adoc (which should be gcode/g-code_fr.adoc) has a very different routine to specify the anchors for each G-code.

@petterreinholdtsen , Is there a chance you rebase against my branch behind https://github.com/LinuxCNC/linuxcnc/pull/1670 "EnglishAnchors" and then try again? That has all files renamed, put files in the right directory, adjusted the master .adoc files and doc.xml and the html files. And, hey, after many tries this now builds. And so should your po4a-motivated changes.

**Comment by petterreinholdtsen** [comment]
[Steffen Möller]
> The file html/gcode_fr causes the errors
> (https://github.com/LinuxCNC/linuxcnc/runs/5265160863?check_suite_focus=true#step:5:7901)
> . This is (other than I initially thought) not autogenerated from the
> .adoc and dependent on the .po files but is edited manually. The file
> has stale references since the gcode/gcode_fr.adoc (which should be
> gcode/g-code_fr.adoc) has a very different routine to specify the
> anchors for each G-code.

Aha.

> @petterreinholdtsen , Is there a chance you rebase against my branch
> behind https://github.com/LinuxCNC/linuxcnc/pull/1670 "EnglishAnchors"
> and then try again? That has all files renamed, put files in the right
> directory, adjusted the master .adoc files and doc.xml and the html
> files. And, hey, after many tries this now builds. And so should your
> po4a-motivated changes.

I'll give it a try.
-- 
Happy hacking
Petter Reinholdtsen

**Comment by petterreinholdtsen** [comment]
[Steffen Möller]
> The file html/gcode_fr causes the errors
> (https://github.com/LinuxCNC/linuxcnc/runs/5265160863?check_suite_focus=true#step:5:7901)
> . This is (other than I initially thought) not autogenerated from the
> .adoc and dependent on the .po files but is edited manually. The file
> has stale references since the gcode/gcode_fr.adoc (which should be
> gcode/g-code_fr.adoc) has a very different routine to specify the
> anchors for each G-code.

I had a look, and this do not make sense.  The html/gcode_fr.html file
should go away with po4a, and show up in its translated state as
docs/html/fr/gcode.html.  Same with gcode/g-code_fr.adoc, which should
show up as docs/src/fr/gcode/g-code.adoc.  Are you basing your tries on
my translate-po4a branch or something else?  What did I misunderstand?

> @petterreinholdtsen , Is there a chance you rebase against my branch
> behind https://github.com/LinuxCNC/linuxcnc/pull/1670 "EnglishAnchors"
> and then try again?

Your branch seem to have more than 40 commits on top of master, with
lots of changes unrelated to the po4a build system, so I prefer to
cherry-pick the relevant parts into mine.

-- 
Happy hacking
Petter Reinholdtsen

**Comment by smoe** [comment]
Maybe a quick fix for you would be as easy as a git rm html/gcode_fr.adoc plus a removal from whereever it is included.

There are so many patches since this continued with the changes branch. This should get merged. And that number of patches got inflated with the review system of github. It is a lot, bot nothing big.

You can imagine that after all the investments I made in beautifying all the different parts of the .adocs to help the .po generation, I do want to see this investment of my time accepted. I do not think that there is much (if anything) in there that was not motivated by the quest to create better .po files, so there should be no point in short-cutting it just to see the po4a build work. The effect would be that there would be less of an incentive to address these two PRs that lead towards it, which would then weaken the po4a effort.

So, please, I strongly prefer that you don't cherry-pick but review what I have done and express that you agree to those changes and hoe that you send out requests for the two PRs to be merged.

**Comment by smoe** [comment]
> Are you basing your tries on my translate-po4a branch or something else? What did I misunderstand?

a) I got the same error messages when I renamed gcode/gcode_fr.adoc o gcode/g-code_fr.adoc without informing (was it html/gcode_fr.html?) about that filename change.
b) The error message of your build on https://github.com/LinuxCNC/linuxcnc/runs/5265160863?check_suite_focus=true#step:5:7901 explicitly points to html/gcode_fr.html so I did not look much further.

**Comment by petterreinholdtsen** [comment]
I am unable to replicate the build errors.  The build work when I test it locally.  Anyone got a clue what is going on here?
